index.bs

```diff
@@ -2028,9 +2039,235 @@ BrowserResult = (
+   window/width=] is equal to the width of the [=web-exposed screen area=] and
+   |window|'s [=client window/height=] is equal to the height of the
+   [=web-exposed screen area=].
+
+1. Return "<code>normal</code>".
+
+</div>
+
+<div algorithm>
+
+To <dfn>set the client window state</dfn> given |window| and |state|:
+
+1. Let |current state| be [=get the client window state| with |window|.
+
+1. If |current state| is equal to |state|, return [=success=] with data null.
+
+1. Switch on the value of |state|:
+
+   <dl>
+    <dt>"<code>fullscreen</code>"
+    <dd>If not [=fullscreen is supported=] return [=error=] with [=error code=]
+    [=unsupported operation=].
+
+    <dt>"<code>normal</code>"
+    <dd>If not [=restore client window is supported=] for |window| return [=error=] with
+    [=error code=] [=unsupported operation=].
+
+    <dt>"<code>maximize</code>"
+    <dd>If not [=maximize client window is supported=] for |window| return [=error=] with
+    [=error code=] [=unsupported operation=].
+
+    <dt>"<code>minimize</code>"
+    <dd>If not [=minimize client window is supported=] for |window| return [=error=] with
+    [=error code=] [=unsupported operation=].
+   </dl>
+
+1. Let |documents| be an empty [=/list=].
+
+1. For each [=/top-level traversable=] |traversable|:
+
+  1. If |traversable|'s associated [=client window=] is not |window| then
+     continue.
+
+  1. Let |document| be |traversable|'s [=active document=].
+
+  1. Append |document| to |documents|.
+
+1. If |documents| is [=list/empty=] return [=error=] with [=error code=]
+   [=no such client window=].
+
+1. If |current state| is "<code>fullscreen</code>":
+
+  1. For each |document| in |documents|:
+
+    1. [=Fully exit fullscreen=] with |document|.
+
+       Note: This is a no-op for documents in window that are not fullscreen.
+
+1. Switch on the value of |state|:
+
+   <dl>
+    <dt>"<code>fullscreen</code>"
+    <dd>
+      1. For each |document| in |documents|:
+
+        1. If |document|'s [=visibility state=] is "<code>visible</code>",
+           [=fullscreen an element=] with |document|'s [=document element=].
+
+        1. [=Break=].
+
+    <dt>"<code>normal</code>"
+    <dd>1. [=Restore the client window=] |window|.
+
+    <dt>"<code>maximize</code>
+    <dd>1. [=Maximize the client window=] |window|.
+
+    <dt>"<code>minimize</code>
+    <dd>1. [=Minimize the client window=] |window|.
+
+1. Return [=success=] with data null.
+
+</div>
 
 ### Types ### {#module-browser-types}
 
+#### The browser.ClientWindow Type #### {#type-browser-ClientWindow}
+
+<pre class="cddl remote-cddl local-cddl">
+browser.ClientWindow = text;
+</pre>
+
+The <code>browser.ClientWindow</code> uniquely identifies a [=client window=].
+
+#### The browser.ClientWindowInfo Type #### {#type-browser-ClientWindowInfo}
+
+<pre class="cddl remote-cddl local-cddl">
+browser.ClientWindowInfo = {
+  clientWindow: browser.ClientWindow,
+  hasFocus: bool,
```

Comment:
We already have the command `browsingContext.activate` which activates (sets the focus to) a tab. Maybe we should stick with `active` here to reduce confusion?

Reply:
+

Reply:
Technically yes, for example during navigation when the document itself isn't fully active.

You can also have the case where the window has focus, but none of the documents do (e.g. when the focus is on the address bar).

Per HTML something can be "have system focus" but not be visible, whereas the `activate` command tries to ensure that the browsing context is actually visible. 

https://html.spec.whatwg.org/#data-model explains the terminology here.

I guess it's OK to use `active` here, but it does mean ignoring some of the edge cases.